spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java

```diff
@@ -139,32 +138,88 @@ public NativeEnvironmentRepository nativeEnvironmentRepository() {
 	@Autowired
 	private ConfigServerProperties server;
 
+	@Autowired(required = false)
+	private SvnKitEnvironmentProperties environmentProperties;
+
 	@Bean
-	public SvnKitEnvironmentRepository svnKitEnvironmentRepository() {
-		SvnKitEnvironmentRepository repository = new SvnKitEnvironmentRepository(
-				this.environment);
-		if (this.server.getDefaultLabel() != null) {
-			repository.setDefaultLabel(this.server.getDefaultLabel());
-		}
-		return repository;
+	public SvnKitEnvironmentRepository svnKitEnvironmentRepository() throws Exception {
+		return new SvnEnvironmentRepositoryFactory(environment, server).build(environmentProperties);
 	}
 }
 
 @Configuration
 @Profile("vault")
 class VaultRepositoryConfiguration {
+	@Autowired(required = false)
+	private VaultEnvironmentProperties environmentProperties;
+
 	@Bean
 	public VaultEnvironmentRepository vaultEnvironmentRepository(
-			HttpServletRequest request, EnvironmentWatch watch) {
-		return new VaultEnvironmentRepository(request, watch, new RestTemplate());
+			HttpServletRequest request, EnvironmentWatch watch) throws Exception {
+		return new VaultEnvironmentRepositoryFactory(request, watch).build(environmentProperties);
 	}
 }
 
 @Configuration
 @Profile("jdbc")
 class JdbcRepositoryConfiguration {
+	@Autowired(required = false)
+	private JdbcEnvironmentProperties environmentProperties;
+
 	@Bean
-	public JdbcEnvironmentRepository jdbcEnvironmentRepository(JdbcTemplate jdbc) {
-		return new JdbcEnvironmentRepository(jdbc);
+	public JdbcEnvironmentRepository jdbcEnvironmentRepository(JdbcTemplate jdbc) throws Exception {
+		return new JdbcEnvironmentRepositoryFactory(jdbc).build(environmentProperties);
 	}
-}
\ No newline at end of file
+}
+
+@Configuration
+@Profile("composite")
+class CompositeRepositoryConfiguration {
+
+	@Bean
+	public Map<String, EnvironmentRepositoryFactory<?,?>> environmentRepositoryFactories(
+			ConfigurableEnvironment environment, ConfigServerProperties server,
+			Optional<TransportConfigCallback> transportConfigCallback, HttpServletRequest request,
+			EnvironmentWatch watch, JdbcTemplate jdbc) {
+		Map<String, EnvironmentRepositoryFactory<?,?>> factories = new HashMap<>();
+		factories.put("git", new GitEnvironmentRepositoryFactory(environment, server, transportConfigCallback));
```

Comment:
I don't think a static list of factories is flexible enough. How would custom repositories work?

Reply:
The `EnvironmentRepositoryFactory`s should be beans that get injected. Spring has a [`FactoryBean<T>`] (https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/beans/factory/config/AbstractFactoryBean.html) that should probably be used. @dsyer would know more about that.

Reply:
@spencergibb I've added dynamic factory resolution and also added a test to show how custom repositories would work.